src/components/shared/thread/ActivityGraphFills2.js

```diff
@@ -0,0 +1,815 @@
+  +mutableFills: CategoryFill[];
+
+  constructor(
+    renderedComponentSettings: RenderedComponentSettings,
+    mutablePercentageBuffers: SelectedPercentageAtPixelBuffers[],
+    mutableFills: CategoryFill[]
+  ) {
+    this.renderedComponentSettings = renderedComponentSettings;
+    this.mutablePercentageBuffers = mutablePercentageBuffers;
+    this.mutableFills = mutableFills;
+  }
+
+  /**
+   * Run the computation to compute a list of the fills that need to be drawn for the
+   * ThreadActivityGraph.
+   */
+  run(): void {
+    // First go through each sample, and set the buffers that contain the percentage
+    // that a category contributes to a given place in the X axis of the chart.
+    this._accumulateSampleCategories();
+
+    debugger;
+
+    // Smooth the graphs by applying a 1D gaussian blur to the per-pixel
+    // contribution of each fill.
+    for (const fill of this.mutableFills) {
+      _applyGaussianBlur1D(fill.perPixelContribution, BOX_BLUR_RADII);
+    }
+
+    this._accumulateUpperEdge();
+  }
+
+  /**
+   * Accumulate the per pixel contribution of each fill, so that each fill's
+   * accumulatedUpperEdge array describes the shape of the "upper edge" after this fill.
+   * Fills are stacked on top of each other.
+   */
+  _accumulateUpperEdge(): void {
+    const { mutableFills } = this;
+    {
+      // Only copy the first array, as there is no accumulation.
+      const { accumulatedUpperEdge, perPixelContribution } = mutableFills[0];
+      for (let i = 0; i < perPixelContribution.length; i++) {
+        accumulatedUpperEdge[i] = perPixelContribution[i];
+      }
+    }
+
+    // Now accumulate the upper edges.
+    let previousUpperEdge = mutableFills[0].accumulatedUpperEdge;
+    for (const {
+      perPixelContribution,
+      accumulatedUpperEdge,
+    } of mutableFills.slice(1)) {
+      for (let i = 0; i < perPixelContribution.length; i++) {
+        accumulatedUpperEdge[i] =
+          previousUpperEdge[i] + perPixelContribution[i];
+      }
+      previousUpperEdge = accumulatedUpperEdge;
+    }
+  }
+
+  /**
+   * Go through each sample, and apply its category percentages to the category
+   * percentage buffers. These percentage buffers determine the overall percentage
+   * that a category contributes to a single pixel. These buffers are mutated in place
+   * with these methods.
+   */
+  _accumulateSampleCategories() {
+    const {
+      fullThread: { samples, stackTable },
+      interval,
+      greyCategoryIndex,
+    } = this.renderedComponentSettings;
+
+    if (samples.length === 0) {
+      // If we have no samples, there's nothing to do.
+      return;
+    }
+
+    let prevSampleTime = samples.time[0] - interval;
+    let sampleTime = samples.time[0];
+
+    // Go through the samples and accumulate the category into the percentageBuffers.
+    for (let i = 0; i < samples.length - 1; i++) {
+      const nextSampleTime = samples.time[i + 1];
+      let sampleCPU;
```

Comment:
I think a better name for this variable (and for the argument of `_accumulateInCategory`) would be `cpuBeforeSample`. And then you can make another variable called `cpuAfterSample`. This will make it clear that we do not know the CPU usage "at" a sample time, we only know it for the interval between samples.